### ISSUE TITLE
Automated Data Science bootstrap from curated content sets

### ISSUE THREAD

**Issue by codificat** [thoth-station/core#402]
## Problem statement

As a Data Scientist,
I want a service that provides me an easy mechanism to bootstrap a new Data Science project, starting from a curated software stack that is appropriate for my project’s goals and available in a shared environment,
so that I can quickly start working on the Data Science project tasks without having to invest time in preparing a working environment, and I can be confident that the project is reproducible and maintainable

## High-level Goals

Starting from an empty git repository, user interacts with bots that populate it from a relevant curated software stack, and make the git project readily available to start working on the Data Science tasks.

This involves:
- A catalog of curated software stacks. Currently we have predictable stacks for [Image Processing](https://github.com/thoth-station/ps-ip/), [Computer Vision](https://github.com/thoth-station/ps-cv) and [Natural Language Processing](https://github.com/thoth-station/ps-nlp))
- A "bootstrap" command for the bot
- Pipelines that create a working build of the project content
- (optional) an online Open Data Hub environment that hosts a running version of the project

## Proposal description

As a Data Scientist I want to open an Issue *"please create an Image Processing notebook"* on GitHub that triggers Thoth bot to start populating my repository:

0. User initializes an empty repo and installs Kebechet
1. User opens an issue in the repo, e.g. "New content set"
2. bot bootstraps repo
3. bot kicks off the Bring-Your-Own-Notebook workflow
4. user enters notebook spawner on ODH@op1st and sees the spawnable notebook image and can start it

### Alternatives

User manually doing each step

## Acceptance Criteria

- [ ] A service entry point / welcome page provides:
  - [ ]  A catalog of curated software stacks. 
  - [ ] Clear and concise instructions on how to use the service
  - [ ] Additional documentation and references to the components involved (Thoth advise, pipelines, byon/odh...)
- [ ] A "bootstrap" command for the bot causes the repo to be pre-populated with the chosen stack content
- [ ] Build pipelines create a working build of the new project content
- [ ] (optional) an online Open Data Hub environment hosts a running version of the project

**Comment by fridex** [comment]
We have discussed an approach that would reuse logic for template projects. Let's sync if we want to develop and maintain this type of logic in Kebechet.

**Comment by durandom** [comment]
>  A "bootstrap" command for the bot causes the repo to be pre-populated with the chosen stack content

I suggest making this the first milestone. It's ok to assume the user knows which stacks are available. So that 
* I start with an empty repo
* I open an issue and ask the bot to create a PR for e.g. the Image Recognition Stack
* I can merge the PR and have a configured repo, which receives update recommendations via PRs in the future

does that make sense?

**Comment by codificat** [comment]
> >  A "bootstrap" command for the bot causes the repo to be pre-populated with the chosen stack content
>
> I suggest making this the first milestone. 

There is a pre-requisite to this, which is that the stack content is readily-usable. This fits with Frido's comment about *template projects*.

> So that
> * I start with an empty repo

.. and install the bot.

> * I open an issue and ask the bot to create a PR for e.g. the Image Recognition  Stack
> * I can merge the PR and have a configured repo,

These 3 steps can be potentially reduced to one by just using the template project logic.

As this is a pre-requisite anyway, I suggest we make that the first milestone. The bot automation can be a follow-up.

I updated the description a bit to reflect that, with the template logic being "phase 1".

Makes sense?

**Comment by MichaelClifford** [comment]
Not sure how you plan to implement this, but sounds like it would require the addition of an ever growing set of template repo's  (is that right?). Have you considered using a cookie-cutter [[1](https://github.com/cookiecutter/cookiecutter )] like repo that would serve as a single repo, with dynamic options that can be implemented on  new repo creations based on the users specific needs?    

You can also look at [[2](https://github.com/aicoe-aiops/cookiecutter-data-science)] our ds cookie cutter repo for a very simple example.  

[1] https://github.com/cookiecutter/cookiecutter   
[2] https://github.com/aicoe-aiops/cookiecutter-data-science

**Comment by codificat** [comment]
> Not sure how you plan to implement this, but sounds like it would require the addition of an ever growing set of template repo's (is that right?). 

Correct, that was the idea: start initially with the 3 "predictable stacks" we have been working on, but eventually offer more options.

> Have you considered using a cookie-cutter [[1](https://github.com/cookiecutter/cookiecutter)] like repo

I saw cookie-cutter and the work you are doing with it, and I was planning to look closer at it, starting with one of the repos (see mention of cookie-cutter/your template in https://github.com/thoth-station/ps-nlp/issues/154 as an option to explore).

But I was still thinking on separate repos. Honestly, this did not occur to me:

> that would serve as a single repo, with dynamic options that can be implemented on new repo creations based on the users specific needs?

Thanks for the suggestion! I will look closer.

An initial question that comes to mind, though: wouldn't that *single repo* become too big/complex?  e.g. the NLP stack alone already has 4 overlays. One goal of this functionality is to be simple, easy to understand - it is meant to bootstrap/get started, and I am wondering if we would potentially be over-complicating the starting point.
